### PR TITLE
History filter

### PR DESCRIPTION
**Pull request by ellie**
Switch between different search modes to narrow down the history you
want - global search for all history, host for all history from your
current machine, session for the current shell session, and directory
for the current directory

The default can be configured via `filter_mode`

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -63,7 +63,7 @@ jobs:
         key: ${{ runner.os }}-cargo-debug-${{ hashFiles('**/Cargo.lock') }}
 
     - name: Run cargo test
-      run: cargo test --workspace
+      run: ATUIN_SESSION=beepboopiamasession cargo test --workspace
```
Comment (conradludgate) [inline .github/workflows/rust.yml:66]: 😂

File: atuin-client/src/database.rs [hunk 2]
```diff
@@ -179,33 +186,58 @@ impl Database for Sqlite {
...
+        unique: bool,
+    ) -> Result<Vec<History>> {
         debug!("listing history");
+        let session = env::var("ATUIN_SESSION")
+            .expect("failed to find ATUIN_SESSION - check your shell setup");
+        let hostname = format!("{}:{}", whoami::hostname(), whoami::username());
+        let cwd = match env::current_dir() {
+            Ok(dir) => dir.display().to_string(),
+            Err(_) => String::from(""),
+        };
```
Comment (conradludgate) [inline atuin-client/src/database.rs:202]: Should this really be in the db layer

**Review by conradludgate — COMMENTED** [review]
Oh jeez that sql

File: .github/workflows/rust.yml (hunk 1 above)
Comment (ellie) [reply .github/workflows/rust.yml:66]: well it's not lying is it? 😏

File: atuin-client/src/database.rs (hunk 2 above)
Comment (ellie) [reply atuin-client/src/database.rs:202]: Meh probably not, but also I don't want to get to the point where we're passing around 9999999 parameters to every database function 🙃

**Comment by ellie** [comment]
This 100% highlights the fact we need to bin off SQL, or at least stop using it in this way. Hopefully a query builder or something would work, I am also not a fan of ORMs 😬

File: atuin-client/src/database.rs (hunk 2 above)
Comment (conradludgate) [reply atuin-client/src/database.rs:202]: I have some thoughts but I can experiment later

**Comment by conradludgate** [comment]
> This 100% highlights the fact we need to bin off SQL, or at least stop using it in this way. Hopefully a query builder or something would work, I am also not a fan of ORMs 😬 

Yeah, I was looking at the sqlx macros for a bit but decided that query builders really is the way to go for us

File: atuin-client/src/database.rs (hunk 2 above)
Comment (ellie) [reply atuin-client/src/database.rs:202]: It's chill, I'm wrapping it up into a struct I think. Those three values are rarely used in isolation, and I should stop being lazy

**Review by conradludgate — APPROVED** [review]

**Review by conradludgate — APPROVED** [review]

**Pull request merged**